turbo/transactions/call.go

```diff
@@ -124,3 +126,99 @@ func getHashGetter(requireCanonical bool, tx kv.Tx, headerReader services.Header
 		return h.Hash()
 	}
 }
+
+type ReusableCaller struct {
+	evm         *vm.EVM
+	gasCap      uint64
+	baseFee     *uint256.Int
+	stateReader state.StateReader
+}
+
+func (r *ReusableCaller) DoCall(
+	ctx context.Context,
+	args ethapi2.CallArgs,
+	callTimeout time.Duration,
+) (*core.ExecutionResult, error) {
+	var cancel context.CancelFunc
+	if callTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, callTimeout)
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+
+	// Make sure the context is cancelled when the call has completed
+	// this makes sure resources are cleaned up.
+	defer cancel()
+
+	msg, err := args.ToMessage(r.gasCap, r.baseFee)
+	if err != nil {
+		return nil, err
+	}
+
+	// reset the EVM so that we can continue to use it with the new context
+	txCtx := core.NewEVMTxContext(msg)
+	r.evm.Reset(txCtx, r.evm.IntraBlockState())
+
+	timedOut := false
+	go func() {
+		<-ctx.Done()
+		timedOut = true
+	}()
```

Comment:
for future PR's - if you will have some idea - how we can avoid spawing this goroutine - welcome

Reply:
for example instead of creating bool variable, can check that ctx.Done() channel is closed

Reply:
Be nice to get rid of this anyway, caused me headaches as the original code was calling `.Cancel()` on the evm which I can't do here, but still want to know about timeouts.  I had a race condition between calls which sometimes caused errors so stuck with the boolean in the end.  Will have a think

Reply:
OK, if we're happy with just knowing 'something' timed out here without trying to cancel the EVM execution we could just use a select statement at the end of the func.  It would terminate the binary search early and stop a coroutine being spawned.  What do you think?

Reply:
let's try in another PR